File: mobile/src/main/java/com/trangiabao/sixjars/base/ui/NumericEditText.java
package com.trangiabao.sixjars.base.ui;

import android.content.Context;
import android.support.v7.widget.AppCompatEditText;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.util.AttributeSet;
import android.view.View;

import java.text.DecimalFormatSymbols;
import java.text.NumberFormat;
import java.text.ParseException;
import java.util.ArrayList;
import java.util.List;

public class NumericEditText extends AppCompatEditText {

    private static final char GROUPING_SEPARATOR = DecimalFormatSymbols.getInstance().getGroupingSeparator();
    private static final char DECIMAL_SEPARATOR = DecimalFormatSymbols.getInstance().getDecimalSeparator();
    private static final String LEADING_ZERO_FILTER_REGEX = "^0+(?!$)";

    private String mDefaultText = null;
    private String mPreviousText = "";
    private String mNumberFilterRegex = "[^\\d\\" + DECIMAL_SEPARATOR + "]";
    private List<NumericValueWatcher> mNumericListeners = new ArrayList<>();

    private final TextWatcher mTextWatcher = new TextWatcher() {
        private boolean validateLock = false;

        @Override
        public void afterTextChanged(Editable s) {
            if (validateLock)
                return;
            if (countMatches(s.toString(), String.valueOf(DECIMAL_SEPARATOR)) > 1) {
                validateLock = true;
                setText(mPreviousText);
                setSelection(mPreviousText.length());
                validateLock = false;
                return;
            }
            if (s.length() == 0) {
                handleNumericValueCleared();
                return;
            }
            setTextInternal(format(s.toString()));
            setSelection(getText().length());
            handleNumericValueChanged();
        }

        @Override
        public void beforeTextChanged(CharSequence s, int start, int count, int after) {
        }

        @Override
        public void onTextChanged(CharSequence s, int start, int before, int count) {
        }
    };

    public NumericEditText(Context context, AttributeSet attrs) {
        super(context, attrs);
        addTextChangedListener(mTextWatcher);
        setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                setSelection(getText().length());
            }
        });
    }

    private void handleNumericValueCleared() {
        mPreviousText = "";
        for (NumericValueWatcher listener : mNumericListeners) {
            listener.onCleared();
        }
    }

    private void handleNumericValueChanged() {
        mPreviousText = getText().toString();
        for (NumericValueWatcher listener : mNumericListeners) {
            listener.onChanged(getNumericValue());
        }
    }

    public void addNumericValueChangedListener(NumericValueWatcher watcher) {
        mNumericListeners.add(watcher);
    }

    public void removeAllNumericValueChangedListeners() {
        while (!mNumericListeners.isEmpty()) {
            mNumericListeners.remove(0);
        }
    }

    public void setDefaultNumericValue(double defaultNumericValue, final String defaultNumericFormat) {
        mDefaultText = String.format(defaultNumericFormat, defaultNumericValue);
        setTextInternal(mDefaultText);
    }

    public void clear() {
        setTextInternal(mDefaultText != null ? mDefaultText : "");
        if (mDefaultText != null) {
            handleNumericValueChanged();
        }
    }

    public double getNumericValue() {
        String original = getText().toString().replaceAll(mNumberFilterRegex, "");
        try {
            return NumberFormat.getInstance().parse(original).doubleValue();
        } catch (ParseException e) {
            return Double.NaN;
        }
    }

    private String format(final String original) {
        final String[] parts = original.split("\\" + DECIMAL_SEPARATOR, -1);
        String number = parts[0].replaceAll(mNumberFilterRegex, "")
                .replaceFirst(LEADING_ZERO_FILTER_REGEX, "");
        number = reverse(reverse(number).replaceAll("(.{3})", "$1" + GROUPING_SEPARATOR));
        number = removeStart(number, String.valueOf(GROUPING_SEPARATOR));
        if (parts.length > 1) {
            number += DECIMAL_SEPARATOR + parts[1];
        }
        return number;
    }

    private void setTextInternal(String text) {
        removeTextChangedListener(mTextWatcher);
        setText(text);
        addTextChangedListener(mTextWatcher);
    }

    private String reverse(String original) {
        if (original == null || original.length() <= 1) {
            return original;
        }
        return TextUtils.getReverse(original, 0, original.length()).toString();
    }

    private String removeStart(String str, String remove) {
        if (TextUtils.isEmpty(str)) {
            return str;
        }
        if (str.startsWith(remove)) {
            return str.substring(remove.length());
        }
        return str;
    }

    private int countMatches(String str, String sub) {
        if (TextUtils.isEmpty(str)) {
            return 0;
        }
        int lastIndex = str.lastIndexOf(sub);
        if (lastIndex < 0) {
            return 0;
        } else {
            return 1 + countMatches(str.substring(0, lastIndex), sub);
        }
    }

    public interface NumericValueWatcher {
        void onChanged(double newValue);

        void onCleared();
    }
}

File: mobile/src/main/java/com/trangiabao/sixjars/ui/NumericEditText.kt
package com.trangiabao.sixjars.ui

import android.content.Context
import android.support.design.widget.TextInputEditText
import android.text.Editable
import android.text.TextUtils
import android.text.TextWatcher
import android.util.AttributeSet
import java.text.DecimalFormatSymbols
import java.text.NumberFormat
import java.text.ParseException
import java.util.*

class NumericEditText(context: Context, attrs: AttributeSet) : TextInputEditText(context, attrs) {

    private var mDefaultText: String? = null
    private var mPreviousText = ""
    private val mNumberFilterRegex = "[^\\d\\$DECIMAL_SEPARATOR]"
    private val mNumericListeners = ArrayList<NumericValueWatcher>()

    private val mTextWatcher = object : TextWatcher {
        private var validateLock = false

        override fun afterTextChanged(s: Editable) {
            if (validateLock)
                return
            if (countMatches(s.toString(), DECIMAL_SEPARATOR.toString()) > 1) {
                validateLock = true
                setText(mPreviousText)
                setSelection(mPreviousText.length)
                validateLock = false
                return
            }
            if (s.isEmpty()) {
                handleNumericValueCleared()
                return
            }
            setTextInternal(format(s.toString()))
            setSelection(text.length)
            handleNumericValueChanged()
        }

        override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}

        override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}
    }

    init {
        addTextChangedListener(mTextWatcher)
        setOnClickListener { setSelection(text.length) }
    }

    private fun handleNumericValueCleared() {
        mPreviousText = ""
        for (listener in mNumericListeners) {
            listener.onCleared()
        }
    }

    private fun handleNumericValueChanged() {
        mPreviousText = text.toString()
        for (listener in mNumericListeners) {
            listener.onChanged(numericValue)
        }
    }

    fun addNumericValueChangedListener(watcher: NumericValueWatcher) {
        mNumericListeners.add(watcher)
    }

    fun removeAllNumericValueChangedListeners() {
        while (!mNumericListeners.isEmpty()) {
            mNumericListeners.removeAt(0)
        }
    }

    fun clear() {
        setTextInternal(mDefaultText!!)
        if (mDefaultText != null) {
            handleNumericValueChanged()
        }
    }

    val numericValue: Double
        get() {
            val original = text.toString().replace(mNumberFilterRegex.toRegex(), "")
            try {
                return NumberFormat.getInstance().parse(original).toDouble()
            } catch (e: ParseException) {
                return java.lang.Double.NaN
            }

        }

    private fun format(original: String): String {
        val parts = original.split(("\\" + DECIMAL_SEPARATOR).toRegex()).toTypedArray()
        var number = parts[0].replace(mNumberFilterRegex.toRegex(), "")
                .replaceFirst(LEADING_ZERO_FILTER_REGEX.toRegex(), "")
        number = reverse(reverse(number).replace("(.{3})".toRegex(), "$1$GROUPING_SEPARATOR"))
        number = removeStart(number, GROUPING_SEPARATOR.toString())
        if (parts.size > 1) {
            number += DECIMAL_SEPARATOR + parts[1]
        }
        return number
    }

    private fun setTextInternal(text: String) {
        removeTextChangedListener(mTextWatcher)
        setText(text)
        addTextChangedListener(mTextWatcher)
    }

    @Suppress("DEPRECATION")
    private fun reverse(original: String?): String {
        if (original == null || original.length <= 1) {
            return original!!
        }
        return TextUtils.getReverse(original, 0, original.length).toString()
    }

    private fun removeStart(str: String, remove: String): String {
        if (TextUtils.isEmpty(str)) {
            return str
        }
        if (str.startsWith(remove)) {
            return str.substring(remove.length)
        }
        return str
    }

    private fun countMatches(str: String, sub: String): Int {
        if (TextUtils.isEmpty(str)) {
            return 0
        }
        val lastIndex = str.lastIndexOf(sub)
        if (lastIndex < 0) {
            return 0
        } else {
            return 1 + countMatches(str.substring(0, lastIndex), sub)
        }
    }

    interface NumericValueWatcher {
        fun onChanged(newValue: Double)

        fun onCleared()
    }

    companion object {
        private val GROUPING_SEPARATOR = DecimalFormatSymbols.getInstance().groupingSeparator
        private val DECIMAL_SEPARATOR = DecimalFormatSymbols.getInstance().decimalSeparator
        private val LEADING_ZERO_FILTER_REGEX = "^0+(?!$)"
    }
}
